File: app/subscriber.py
#!/usr/bin/env python3

import json
import sys
import uuid
import base64
import argparse
import subprocess
from websocket import WebSocketApp
import boto3
from botocore.auth import SigV4Auth
from botocore.awsrequest import AWSRequest
from botocore.credentials import Credentials
from urllib.parse import urlparse
from termcolor import colored, cprint

kaCount = 0

DEFAULT_HEADERS = {
    'accept': 'application/json, text/javascript',
    'content-encoding': 'amz-1.0',
    'content-type': 'application/json; charset=UTF-8',
}

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('--api-id', required=True)
    parser.add_argument('--channel', default='/default/*')
    parser.add_argument('--region', default=None)
    return parser.parse_args()

def get_api(api_id):
    try:
        appsync = my_session.client('appsync')
        response = appsync.get_api(apiId=api_id)
        return response['api']
    except Exception as e:
        print(f"Error: {str(e)}")
        sys.exit(1)

def get_session_token():
    sts = my_session.client('sts')
    return sts.get_session_token(DurationSeconds=3600)

def sign(api, credentials, body=None):
    url = f"https://{api['dns']['HTTP']}/event"
    request = AWSRequest(
        method='POST',
        url=url,
        data=body if body else '{}',
        headers=DEFAULT_HEADERS
    )

    auth = SigV4Auth(credentials, 'appsync', my_session.region_name)
    auth.add_auth(request)
    signed = {'host': urlparse(url).netloc}
    signed.update(dict(request.headers))
    return signed

def get_auth_protocol(api, credentials, body=None):
    signed = sign(api, credentials, body)
    json_str = json.dumps(signed)
    b64 = base64.b64encode(json_str.encode()).decode()
    header = b64.replace('+', '-').replace('/', '_').rstrip('=')
    return f"header-{header}"

def on_message(ws, received):
    global kaCount
    message = json.loads(received)

    if message['type'] == 'data':
        data = json.loads(message['event'])
        if kaCount > 0: print("")
        kaCount = 0
        print(f"{colored(">>", "blue", attrs=["bold"])} {data}")
    elif message['type'] == 'ka':
        kaCount = kaCount + 1
        print(
            colored(f">> KA{" (x" +str(kaCount)+ ")" if kaCount > 1 else ""}", "green",attrs=["bold"]),
            end='\r', flush=True)
    else:
        if kaCount > 0: print("")
        kaCount = 0
        print(f">> {message}")

def on_error(ws, error):
    print(f"Error: {error}")

def on_close(ws, close_status_code, close_msg):
    print("Connection closed")

def connectAndSubscribe(ws, credentials, channel):
    ws.send(json.dumps({'type': 'connection_init'}))
    subscribe_msg = {
        'type': 'subscribe',
        'id': str(uuid.uuid4()),
        'channel': args.channel
    }
    print(f"<< {subscribe_msg}")
    auth = sign(api, credentials, json.dumps({'channel': args.channel},separators=(',', ':')))
    subscribe_msg.update({'authorization': auth})

    ws.send(json.dumps(subscribe_msg,separators=(',', ':')))

if __name__ == "__main__":
    args = parse_args()

    my_session = boto3.session.Session(region_name=args.region)

    api = get_api(args.api_id)
    tokens = get_session_token()

    creds = tokens["Credentials"]
    credentials = Credentials(creds['AccessKeyId'], creds['SecretAccessKey'], creds['SessionToken'])
    
    ws = WebSocketApp(
        f"wss://{api['dns']['REALTIME']}/event/realtime",
        header=DEFAULT_HEADERS,
        subprotocols=['aws-appsync-event-ws', get_auth_protocol(api, credentials)],
        on_open=lambda ws: connectAndSubscribe(ws, credentials, args.channel),
        on_message=on_message,
        on_error=on_error,
        on_close=on_close
    )
    
    ws.run_forever()

File: app/subscriber.mjs
#!/usr/bin/env zx

import { AwsV4Signer } from 'aws4fetch'
import { WebSocket } from 'ws'

const DEFAULT_HEADERS = {
  accept: 'application/json, text/javascript',
  'content-encoding': 'amz-1.0',
  'content-type': 'application/json; charset=UTF-8',
}

// Retrieve the api-id and the channel from the command line
const argv = minimist(process.argv.slice(2), { string: ['api-id', 'channel'] })

if (!argv['api-id']) {
  console.log(chalk.red('Usage: subscribe --api-id <id> [--channel <path>]'))
  process.exit(1)
}

const channel = argv.channel || '/default/*'

// Fetch the api using the AWS ClI.
let api
try {
  const result = await $`aws appsync get-api --api-id ${argv['api-id']}`.quiet().json()
  api = result.api
} catch (error) {
  console.log(chalk.white.bgRed(error))
  process.exit(1)
}

// Get temporary tokens using STS
const tokens = await $`aws sts get-session-token`.json()
const auth = await getAuthProtocol(api)

// Connect to the WebSocket
const socket = await new Promise((resolve, reject) => {
  const socket = new WebSocket(
    `wss://${api.dns.REALTIME}/event/realtime`,
    ['aws-appsync-event-ws', auth],
    { headers: { ...DEFAULT_HEADERS } },
  )

  socket.onopen = () => {
    socket.send(JSON.stringify({ type: 'connection_init' }))
    resolve(socket)
  }
  socket.onclose = (evt) => reject(new Error(evt.reason))
  socket.onmessage = (event) => console.log(chalk.blue.bold('>>'), JSON.parse(event.data))
})

// Create th esubscription request and send
const subscribeMsg = { type: 'subscribe', id: crypto.randomUUID(), channel }
console.log(chalk.blue.bold('<<'), subscribeMsg)
socket?.send(JSON.stringify({ ...subscribeMsg, authorization: await sign(api, { channel }) }))

// -- signing functions --

/**
 * @param {any} api  the AppSync API
 * @param {any} [body] the body
 * @returns {}
 */
async function sign(api, body) {
  const signer = new AwsV4Signer({
    url: `https://${api.dns.HTTP}/event`,

    accessKeyId: tokens.Credentials.AccessKeyId,
    secretAccessKey: tokens.Credentials.SecretAccessKey,
    sessionToken: tokens.Credentials.SessionToken,

    method: 'POST',
    headers: DEFAULT_HEADERS,
    body: body ? JSON.stringify(body) : '{}',

    service: 'appsync',
  })

  // Retrieve the headers and host from the signed object
  const { headers, url } = await signer.sign()
  const signed = { host: url.host }
  for (const [k, v] of headers) {
    signed[k] = v
  }

  return signed
}

async function getBase64URLEncoded(api, body) {
  return btoa(JSON.stringify(await sign(api, body)))
    .replace(/\+/g, '-') // Convert '+' to '-'
    .replace(/\//g, '_') // Convert '/' to '_'
    .replace(/=+$/, '') // Remove padding `=`
}

/**
 * @param {any} api the appsync api
 * @param {Record} [body]
 * @returns string
 */
async function getAuthProtocol(api, body) {
  const header = await getBase64URLEncoded(api, body)
  return `header-${header}`
}
